### DISCUSSION TITLE
Dashboard Improvement 1: GPU Widget

### DISCUSSION THREAD

**Discussion by dszidi** [truenas/webui#9865] [Design Previews]
Greetings community members! We have had a lot of feedback from users requesting a GPU widget. We've heard you loud and clear, and would like to present to you some wireframes. 

> Wireframes Disclaimer:
Designs in blue are wireframes that are meant to be low fidelity sketches. They will not be pixel perfect and may not be as visually precise as the final product in the presentation. It may also make use of placeholders for things like images and graphs. Please focus on the content and treat them as drafts.

### About the Designs

The following are designs for what the widget will contain for Nvidia and AMD graphics cards. Due to some backend limitations there will be no Intel support at first until some dependencies get support for those devices. Also due to dependencies being used, not all the same stats are available between Nvidia and AMD. With that said, please take a look at the designs and give us your feedback.

Thanks!

<img width="773" alt="image" src="https://github.com/truenas/webui/assets/29865743/ebeb1c61-06ab-4893-9cea-4793c8d33a03">

**Comment by GMAzrael** [comment]
Looks excellent to me!

**Comment by xtj7** [comment]
A few thoughts:
- I assume with multiple GPUs we simply have one widget per GPU, even if these are multiple Nvidia GPUs for example?
- Has there been thought about a mobile viewport? I realise that probably isn't the highest priority, but checking stats on mobile for troubleshooting can be quite helpful. So especially the dashboards should work well with mobile too.
- Historical data would be a nice touch as well. Maybe configurable with a drop in resolution, similarly to how homeassistant is handling historical data.
- Custom alerts set for reaching a certain GPU usage, VRAM usage or encoder/decoder usage - if my CCTV isn't recording properly because of decoding/encoding limitations, I want to know :) And this could be visualised in the UI as well (for example: make the metric red once it is past the custom alert, so you can see at a single glance where you are above the threshold)
- Custom label or description per GPU: I have one GPU dedicated for my NVR and one GPU for everything else. Knowing which one's stats I'm looking at would be very helpful

**Reply by dszidi** [reply]
>I assume with multiple GPUs we simply have one widget per GPU, even if these are multiple Nvidia GPUs for example?

Correct. If you have three Nvidia cards on your system you will have three GPU widgets on the dashboard

>Historical data would be a nice touch as well. Maybe configurable with a drop in resolution, similarly to how homeassistant is handling historical data.

We treat the dashboard as the place to get information about the current state of the system, and the reporting page as the place to find historical charts and graphs. Typically we keep an icon button on most of our dashboard widgets linking to the relevant reporting page. You can expect to see that on these GPU widgets as well. I will update the designs to reflect that. 

>Custom alerts set for reaching a certain GPU usage, VRAM usage or encoder/decoder usage - if my CCTV isn't recording properly because of decoding/encoding limitations, I want to know :) And this could be visualised in the UI as well (for example: make the metric red once it is past the custom alert, so you can see at a single glance where you are above the threshold)

This sounds interesting.  We'll take it under consideration.

>Custom label or description per GPU: I have one GPU dedicated for my NVR and one GPU for everything else. Knowing which one's stats I'm looking at would be very helpful

This is a great suggestion. I'll investigate how we can incorporate that into our existing designs and update the initial post.

**Comment by Sophist-UK** [comment]
Nice.

**Comment by ElectronicBattle** [comment]
This is an exciting development for me, seeing as (1) I've just fitted a graphics card for use in TN Scale official apps and (2) I'm very much a beginner, having been running TN Scale for only a couple of years as a hobbyist  with no programming background so frequently out of my depth!

**Comment by panzerscope** [comment]
For someone who is looking for a GPU widget on TN Scale, this looks encouraging, although one would have to decide what “GPU Usage” is actually monitoring on the widget. Would it take into account overall usage, or just that of the 3D Engine, or Media Engine for things like transcoding?

The widgets as designed on the old topic you linked look pretty, but IMHO not granular enough. For example The Intel_GPU_Top (for Intel GPU's) gives you a good breakdown of what part of the GPU is being used and by what process, such as the below:

![image](https://github.com/user-attachments/assets/b72b162c-e4bc-462e-b947-693dc6389e20)


I know right now that Intel GPU support is out of scope due to other dependency requirements (hope you guys can solve that) but I would think that having more granular visibility for Nvidia and AMD GPU's would also be helpful.

P.S The last comment on this was July, has there been any progress on this in general?

**Comment by boimouseorange** [comment]


**Reply by Sophist-UK** [reply]
> This seems to be a good thread for me to jump on due to my current issue.

No - this is ***NOT*** a good thread for you to jump on.

1. It is (generally speaking) considered inappropriate to hi-jack one thread to discuss something different.
2. This is a discussion about something completely different, so not many people will latch on to your issue.
3. The correct place for this sort of question is in the TrueNAS forums.

**Comment by PeterKse** [comment]
@dszidi Hi Damian, I'm very keen on this idea of a GPU widget, since it's been long is this something that's still being considered and working on?

**Comment by scyto** [comment]
Love them, when can we have them for real ;-)

**Comment by aindriu80** [comment]
great widget and nice wireframes... hopefully something will come up

**Comment by RandomPedanticTribble** [comment]
_Providing a gentle nudge to indicate I also really really want/need this widget..._

**Comment by Trevo525** [comment]
They look great! has there been any sort of push to get these going since this has been posted?

**Reply by RandomPedanticTribble** [reply]
I had the opportunity to speak to a lovely lady from the Truenas management team who was quite interested in my use case. (Stable diffusion on VMs and GPU's, where the models are shared on Truenas between these VMs and my desktop to have uniform possibilities). And because it's a server and a passivly cooled Tesla GPU, I need to find a good way to watch the temps...

**Comment by alin180493** [comment]
+1 on the GPU Widget

**Comment by mntbighker** [comment]
+1 on the GPU widget

**Comment by NoahsArch** [comment]
This is a no brainer add it!

**Comment by DarthJahus** [comment]
We need this!

**Comment by camargomau** [comment]
I was surprised to find out this widget isn't yet implemented. +1

**Comment by ThisBytes5** [comment]
I want this, are there still plans to add it?

**Comment by CodePunk-** [comment]
Any updates on this?

**Comment by 0310njs** [comment]
is their any other way to track if this is being worked on? Or a way to bump this so it gets more attention?

**Comment by GreatScotlandAU** [comment]
This would be such a good widget to have. Please don't forget about Intel GPUs either

**Comment by halloulpsx** [comment]
Does anyone know if this is still planned?

**Comment by aindriu80** [comment]
i'm sure something could be vibe coded

**Comment by divijshahh** [comment]
we need this done!!!

**Comment by aindriu80** [comment]
I wanted to add some technical context here since the discussion has mostly been +1s without much analysis of what's actually involved. Hopefully this is useful to someone.

**Where does the work actually sit?**

A GPU widget has two distinct parts:

**1. Backend (middleware) - the likely blocker**
The frontend already knows how to display widgets - `WidgetCpuComponent`, `WidgetMemoryComponent` etc. all follow a well-established pattern, subscribing to WebSocket API topics and rendering with the existing chart libraries. A `WidgetGpuComponent` would slot in cleanly alongside these.

The missing piece is a middleware API method (something like `reporting.gpu_stats`) that calls `nvidia-smi --query-gpu` for NVIDIA and `rocm-smi` for AMD, and exposes the result as a real-time WebSocket subscription - the same way CPU metrics are streamed to the dashboard today. The `device.get_gpus` private method exists and already enumerates hardware, but that's static hardware info, not live runtime stats (utilisation %, VRAM used, temperature, encoder/decoder load etc.).

This middleware layer appears to be where progress has stalled, which also aligns with the note in the original post about backend limitations affecting Intel support.

**2. Frontend (Angular/TypeScript) - community-ready**
The webui repo is fully open and the widget pattern is easy to follow. If the middleware endpoint existed, a community contributor could likely build the Angular component and submit a PR without much friction.

**A question**
Is the blocker specifically the middleware stats API? If that endpoint were defined and documented - even as a draft spec - it would allow community contributors to work on the frontend, or potentially contribute the middleware piece too.

The designs from @dszidi look great.

**Comment by FxMorin** [comment]
I would also like this. I've written a draft for the required endpoint.
It would need to be looked at by TrueNas devs to be properly fleshed out.
https://github.com/truenas/middleware/pull/18883

The web part should be pretty easy once thats done.

**Comment by DarthJahus** [comment]
Closed as resolved?

**Reply by 0310njs** [reply]
I have the same question. was the request moved to a higher design phase?

**Comment by 0310njs** [comment]
I think this was closed because they got the feedback they wanted and their is no need to beat the dead horse.

[https://github.com/truenas/webui/discussions/9867](url)
[https://github.com/truenas/webui/discussions/9866](url)

these other discussions have the same GPU wigets.
